src/Compiler/Checking/TypeRelations.fs

```diff
@@ -54,80 +60,111 @@ let stripAll stripMeasures g ty =
+    else
+        ValueNone
+
+let inline UpdateCachedTypeSubsumption (g: TcGlobals) (amap: ImportMap) key subsumes : unit =
+    if g.compilationMode = CompilationMode.OneOff && g.langVersion.SupportsFeature LanguageFeature.UseTypeSubsumptionCache then
+        amap.TypeSubsumptionCache[key] <- subsumes
+
 /// The feasible coercion relation. Part of the language spec.
-let rec TypeFeasiblySubsumesType ndeep g amap m ty1 canCoerce ty2 = 
-    if ndeep > 100 then error(InternalError("recursive class hierarchy (detected in TypeFeasiblySubsumesType), ty1 = " + (DebugPrint.showType ty1), m))
-    let ty1 = stripTyEqns g ty1
-    let ty2 = stripTyEqns g ty2
-    match ty1, ty2 with 
-    | TType_var _, _  | _, TType_var _ -> true
+let rec TypeFeasiblySubsumesType ndeep (g: TcGlobals) (amap: ImportMap) m (ty1: TType) (canCoerce: CanCoerce) (ty2: TType) =
 
-    | TType_app (tc1, l1, _), TType_app (tc2, l2, _) when tyconRefEq g tc1 tc2 ->
-        List.lengthsEqAndForall2 (TypesFeasiblyEquiv ndeep g amap m) l1 l2
+    if ndeep > 100 then
+        error(InternalError("Large class hierarchy (possibly recursive, detected in TypeFeasiblySubsumesType), ty1 = " + (DebugPrint.showType ty1), m))
 
-    | TType_tuple _, TType_tuple _
-    | TType_anon _, TType_anon _
-    | TType_fun _, TType_fun _ ->
-        TypesFeasiblyEquiv ndeep g amap m ty1 ty2
+    let ty1 = stripTyEqns g ty1
+    let ty2 = stripTyEqns g ty2
 
-    | TType_measure _, TType_measure _ ->
-        true
+    // Check if language feature supported
+    let key = TTypeCacheKey (ty1, ty2, canCoerce, g)
+
+    match TryGetCachedTypeSubsumption g amap key with
+    | ValueSome subsumes ->
+        subsumes
+    | ValueNone ->
+        let subsumes =
+            match ty1, ty2 with
+            | TType_measure _, TType_measure _
+            | TType_var _, _ | _, TType_var _ ->
+                true
+
+            | TType_app (tc1, l1, _), TType_app (tc2, l2, _) when tyconRefEq g tc1 tc2 ->
+                List.lengthsEqAndForall2 (TypesFeasiblyEquiv ndeep g amap m) l1 l2
+
+            | TType_tuple _, TType_tuple _
+            | TType_anon _, TType_anon _
+            | TType_fun _, TType_fun _ ->
+                TypesFeasiblyEquiv ndeep g amap m ty1 ty2
+
+            | _ ->
+                // F# reference types are subtypes of type 'obj'
+                    if isObjTy g ty1 && (canCoerce = CanCoerce || isRefTy g ty2) then
+                        true
+                    elif isAppTy g ty2 && (canCoerce = CanCoerce || isRefTy g ty2) && TypeFeasiblySubsumesTypeWithSupertypeCheck g amap m ndeep ty1 ty2 then
+                        true
+                    else
+                        let interfaces = GetImmediateInterfacesOfType SkipUnrefInterfaces.Yes g amap m ty2
+                        // See if any interface in type hierarchy of ty2 is a supertype of ty1
+                        Seq.exists (TypeFeasiblySubsumesType (ndeep + 1) g amap m ty1 NoCoerce) interfaces
```

Comment:
List.exists stops evaluation after first `true` hit as well, so this can stay being a list (and therefore allow further consumers to used inlined List module functions on it)

Reply:
I have rewritten `GetImmediateInterfacesOfType` to return `seq` so we do not allocate the whole list and the throw it away immediately after, but removed the change to minimize the PR. Forgot to revert this place.